specs/_features/eip7805/fork-choice.md

```diff
@@ -0,0 +1,109 @@
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+
+- [Introduction](#introduction)
+- [Helpers](#helpers)
+  - [New `validate_inclusion_lists`](#new-validate_inclusion_lists)
+  - [Modified `Store`](#modified-store)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+<!-- /TOC -->
+
+## Introduction
+
+This is the modification of the fork choice accompanying the EIP-7805 upgrade.
+
+## Configuration
+
+| Name | Value | Unit | Duration |
+| - | - | :-: | :-: |
+| `VIEW_FREEZE_DEADLINE` | `uint64(9)` | seconds | 9 seconds | # [New in EIP-7805]
+
+## Helpers
+
+### New `validate_inclusion_lists`
+
+```python
+def validate_inclusion_lists(store: Store, inclusion_list_transactions: List[Transaction, MAX_TRANSACTIONS_PER_INCLUSION_LIST * IL_COMMITTEE_SIZE], execution_payload: ExecutionPayload) -> bool:
+    """
+    Return ``True`` if and only if the input ``inclusion_list_transactions`` satifies validation, that to verify if the `execution_payload` satisfies `inclusion_list_transactions` validity conditions either when all transactions are present in payload or when any missing transactions are found to be invalid when appended to the end of the payload unless the block is full.
+    """
+    ...
+```
+
+### Modified `Store` 
+**Note:** `Store` is modified to track the seen inclusion lists.
+
+```python
+@dataclass
+class Store(object):
+    time: uint64
+    genesis_time: uint64
+    justified_checkpoint: Checkpoint
+    finalized_checkpoint: Checkpoint
+    unrealized_justified_checkpoint: Checkpoint
+    unrealized_finalized_checkpoint: Checkpoint
+    proposer_boost_root: Root
+    equivocating_indices: Set[ValidatorIndex]
+    blocks: Dict[Root, BeaconBlock] = field(default_factory=dict)
+    block_states: Dict[Root, BeaconState] = field(default_factory=dict)
+    block_timeliness: Dict[Root, boolean] = field(default_factory=dict)
+    checkpoint_states: Dict[Checkpoint, BeaconState] = field(default_factory=dict)
+    latest_messages: Dict[ValidatorIndex, LatestMessage] = field(default_factory=dict)
+    unrealized_justifications: Dict[Root, Checkpoint] = field(default_factory=dict)
+    inclusion_lists: Dict[Tuple[Slot, Root], List[InclusionList]] = field(default_factory=dict) # [New in EIP-7805]
+    inclusion_list_equivocators: Dict[Tuple[Slot, Root], Set[ValidatorIndex]] = field(default_factory=dict) # [New in EIP-7805]
+```
+
+### New `on_inclusion_list`
+
+`on_inclusion_list` is called to import `signed_inclusion_list` to the fork choice store.
+```python
+def on_inclusion_list(
+        store: Store, 
+        signed_inclusion_list: SignedInclusionList, 
+        inclusion_list_committee: Vector[ValidatorIndex, IL_COMMITTEE_SIZE]]) -> None:
+    """
+    ``on_inclusion_list`` verify the inclusion list before importing it to fork choice store.
+    If there exists more than 1 inclusion list in store with the same slot and validator index, remove the original one.
```

Comment:
I believe this is an out-of-date comment. Where is "remove the original one" happening exactly? It appears that it add the validator to `inclusion_list_equivocators` instead now.

Reply:
> I believe this is an out-of-date comment. Where is "remove the original one" happening exactly? It appears that it adds the validator to `inclusion_list_equivocators` instead now.

I thought it's the other way around? 

p2p only accepts an IL only if it's the first or second IL broadcasted by the peer. 

The responsibility of picking out the equivocators should be on the p2p side. Beacon node just replace the first IL with the second IL if that happens

Reply:
The comments are outdated, and different client implementations may handle this differently. Essentially, we want to:

- Cache at least one inclusion list
- Re-broadcast up to two inclusion lists over P2P
- Track which inclusion lists have been equivocated (e.g., identify the specific validator)
- When calling the EL to validate inclusion lists, ignore those that have been equivocated (discard them from being sent to the EL)
- Regarding whether to prioritize the first or second inclusion list, I'm leaning towards the first as it seems simpler

Reply:
The comment is outdated, yeah. It doesn't really matter which one we keep, we could also not keep either since anyway we ignore them once we know that it is an equivocation. For spec simplicity it seemed easier to just not do anything instead of removing, and also perhaps it makes sense to keep the first, just because if you're the builder you would probably rather use it even if there's an equivocation, just in case, and you'd rather use the first since you saw it earlier and from your perspective it is more likely to be enforced by someone